fixes for qt6

ijgnd opened the pull request:
this commit has some fixes for pyqt6 (so that Anki starts without error messages).

For real compatibility to 2.1.50 the changes in Anki's functions would have to be copied over into this add-on. I'll look into this later.

you should test this pull request before releasing it.

Commit:
fixes for qt6

NameLessGO commented:
Fantastic work ijdnd, the only problem found was  "Toggle sidebar" in Qt6, once pressed it does not hide/unhide sidebar (the terminal didn't displayed anything to post here)

fond this on the terminal
```
addons21/46611790/fastbar.py:42:intTime is deprecated: please use 'int_time'
Starting main loop...
addons21/1052724801/ui_browser.py:1:isMac is deprecated: please use 'is_mac'
addons21/1052724801/ui_browser.py:1:pointVersion is deprecated: please use 'point_version'
addons21/1052724801/dialog__multi_line.py:1:isMac is deprecated: please use 'is_mac'
addons21/1052724801/dialog__multi_line.py:1:pointVersion is deprecated: please use 'point_version'
addons21/1052724801/onTextChange.py:1:pointVersion is deprecated: please use 'point_version'
addons21/1052724801/helpers.py:1:pointVersion is deprecated: please use 'point_version'
addons21/1052724801/help_text.py:1:pointVersion is deprecated: please use 'point_version'
addons21/1052724801/filter_button.py:11:pointVersion is deprecated: please use 'point_version'
addons21/1052724801/ui_browser_ComboReplacer.py:44:pointVersion is deprecated: please use 'point_version'
addons21/1052724801/ui_browser_modify_searchEdit.py:2:pointVersion is deprecated: please use 'point_version'
addons21/1052724801/ui_filtered_decks_dialog.py:3:pointVersion is deprecated: please use 'point_version'
```
wait, it is basically to switch up the deprecated to the one it is displaying? if it is then 
```
addons21/46611790/fastbar.py:42:int_time
Starting main loop...
addons21/1052724801/ui_browser.py:1:is_mac
addons21/1052724801/ui_browser.py:1:point_version
addons21/1052724801/dialog__multi_line.py:1:is_mac
addons21/1052724801/dialog__multi_line.py:1:point_version
addons21/1052724801/onTextChange.py:1:point_version
addons21/1052724801/helpers.py:1:point_version
addons21/1052724801/help_text.py:1:point_version
addons21/1052724801/filter_button.py:11:point_version
addons21/1052724801/ui_browser_ComboReplacer.py:44:point_version
addons21/1052724801/ui_browser_modify_searchEdit.py:2:point_version
addons21/1052724801/ui_filtered_decks_dialog.py:3:point_version
```
(after watched the latest LinusTechTIps vid, discovered how amazing is Git and GitHub)

Commit:
fix toggle sidebar for qt6

Commit:
bury/unbury for .45+ for new undo

ijgnd commented:
the commit 9d1e47e makes the sidebar toggle work for builds with qt6. The original approach was not deprecated and should have worked as far as I see. I changed the sidebar toggle code to the approach I use in my [Browser Maximize/Hide Table/Editor](https://ankiweb.net/shared/info/1819291495).

&nbsp;

The commit 297c514 needs testing on .45+. Also check if sync still works (I haven't tested this). I tried to use the new .45+ code so that multiple undo steps are possible.

ijgnd commented:
@NameLessGO: 

- Thanks for your report. Add-on creators benefit a lot from people who give exact bug reports. You can help a lot by testing and creating new issues.
- I haven't fixed all the warnings shown on the terminal because at the moment these are still only warnings which will only break in future versions. But your idea is correct, in the source code I'd have to replce e.g. `isMac` with `is_mac`. If you're curious about why this happened and have started learning python [this](https://github.com/ankitects/anki/pull/1451) is the reason why the changes are needed.
-  the problem is that the new function names only work in recent anki versions. On Ankiweb you can upload different add-on versions for different add-ons but often it's less work to have the code for all anki versions in just one add-on version/file. That's why I often have stuff like [this](https://github.com/ijgnd/Fastbar-with-nightmode-support/blob/297c5146d49088dc4c93b30bffafaca08f4a02e2/addon/fastbar.py#L43) in my files. Anki has a function pointVersion that was recently renamed to point_version but this was only introduced around Anki 2.1.22 and some people still user older anki versions and fastbar is also for those people. That's why I use custom code to identify the anki version used.
- If you want to track down errors in add-ons running anki from the command line and then checking the terminal output is very useful. It's even more useful if you know some python and then insert some print statements into the source code to get info about the values of variables etc. E.g. have a look at [this line](https://github.com/ijgnd/Fastbar-with-nightmode-support/blob/297c5146d49088dc4c93b30bffafaca08f4a02e2/addon/fastbar.py#L197). If you don't know the output of self.selectedCards() and you're not sure after reading the anki source code you could add a line after this like `    print(c)` so that whenever Anki runs this function (which it does when you click the toggle bury button) it would print the value of c to the console output. Also it's often very useful to get the attributes of a variable with a command like `dir(my_var)`, see [here](https://docs.python.org/3/library/functions.html#dir)).
- a lot has changed in the last two years but some stuff might still be relevant [here](https://www.reddit.com/r/Anki/comments/ctmwmn/fastest_and_easiest_way_to_grasp_python_for/) if you want to know more about learning python for anki. As I said there you can do a lot with some basic knowledge so that you can fix broken old add-ons or combine other people's code.
- I know little coding so if you want to look for best practice stuff look at other code ... I just need to tweak Anki ...

NameLessGO commented:
everything worked fine about undo ( but the "edit tags" have some problems
1. about multiple undo steps: didn't undo when add, changed, deleted tags (all inside  "edit tags", instead showed "nothing to undo"

2. "edit tags" shortcut have the same shortcut for "flag6" (this affects .45+ because was when the new flags were added)
 `Qt warning: QAction::event: Ambiguous shortcut overload: Ctrl+6`

 3.  if you do  'search" (inside the "edit tags" options), it adds the tag that was searched into your cards (example: if you have a card without tags, then search for "testtest", anki will search "testtest" as normal, but if you return to the card, there will be added "testtest"
 
 about the sync (shortcut Y), well it worked, didn't knew what was meant to be, so was overthinking for a while lol
 
 decided to watch introduction to [programming](https://youtu.be/zOjov-2OZ0E) before jumping to the 4 hours [python](https://youtu.be/rfscVS0vtbw) course

AnKingMed commented:
I have not had time to test this and won't till the weekend. @NameLessGO is this working well enough to publish yet?

@ijgnd is this intended to publish as a separate branch or is it backwards compatible?

NameLessGO commented:
yup, it is working great, ijgnd nailed it

Commit:
support for all anki versions (since ~2.1.20) in one file

AnKingMed commented:
Awesome. Thank you @ijgnd and @NameLessGO . I'll upload it to ankiweb now

ijgnd commented:
@AnKingMed: This was originally intended for a separate branch, but I just added a commit that should make this version of the add-on work on Anki 2.1.22+. Fastbar nominally works on all anki versions since 2.1.0 but it has been broken on old Anki versions before ~2.1.22 for months (see [here](https://github.com/ankipalace/Fastbar-with-nightmode-support/issues/13)) so I guess that's enough.

I

ijgnd commented:
@AnKingMed: Let's hope that my new unbury code works well ....

&nbsp;

@NameLessGO: Thanks for all your help. Just to be sure you could once more test the unbury function of the fastbar in 2.1.45+, after burying and unburying some cards try syncing to your phone and run check database if something turns up. I don't think so. But I prefer third party confirmation.

1. I'm not sure what you mean. Could you post a slightly longer explanation in a new issue at https://github.com/ijgnd/anki__extended_tag_edit_add_dialog/issues so that our messages are discoverable. This remineded me that this would be nice: https://github.com/ijgnd/anki__extended_tag_edit_add_dialog/issues/8 But I think you mean something different. 
2. "Ctrl+6" is the shortcut AnkingMed used to recommend before Anki started using this simple to type key combo. My default shortcuts was something like "Alt+Shift+6", in mid 2020 I changed to "Ctrl+t, a". I added a warning on top of https://ankiweb.net/shared/info/1135507717. My new two-part default shortcut breaks Anki's two-part shortcuts for Latex unfortunately. [This](https://github.com/ijgnd/anki__extended_tag_edit_add_dialog/commit/fe8836e64387494a700ba0938903bbbc5fc60499) commit should fix the problem in 2.1.50+ but I'm still waiting for feedback from Damien [here](https://forums.ankiweb.net/t/emacs-style-shortcuts-with-ctrl-t-in-the-editor-in-45/12280/10).
3. added this to my issues list as https://github.com/ijgnd/anki__extended_tag_edit_add_dialog/issues/9

> decided to watch introduction to programming before jumping to the 4 hours python course

that's great news. I guess you already use Anki for this - if not [this great essay](http://augmentingcognition.com/ltm.html) might be good motivation.

NameLessGO commented:
experiments with bury/unbury
1. bury all the cards, then select half and unbury -> sync -> check data base on  phone -> sync -> no error
2.  unbury all the cards, then select half and bury -> sync -> check data base on  phone -> sync -> no error
3.  bury all the cards, then select half and unbury -> sync -> check data base on desktop -> sync -> no error
4. unbury all the cards, then select half and bury -> sync -> check data base on desktop -> sync -> no error

1. will test with Extended Tag Add/Edit Dialog then report later
2.  idk why i had typed the crtl + 6 command problem here since it is related to anking configuration
3.  thanks

will read the essay too

ijgnd commented:
@NameLessGO: Luckily just now Damien also had a look at it and said to integrate this function into 2.1.50, see [here](https://forums.ankiweb.net/t/add-a-collection-operation-to-unbury-cards/15720). But I still feel better that you did another test a few hours earlier. A few hours is enough to send the update to a few hundred users .... So thanks once more.

AnKingMed commented:
@NameLessGO  if you have a chance, can you test the new PR that @ijgnd shared?   I won't have time for at least a week unfortunately
